### ISSUE TITLE
CI: Cilium IPsec upgrade (ci-ipsec-upgrade): Setup & Test (2, 5.10-20231026.065108, iptables, disabled, disabled, ipsec, true) [v1.14]

### ISSUE THREAD

**Issue by marseel** [cilium/cilium#29351]
## CI failure

Cilium IPsec upgrade (ci-ipsec-upgrade) is persistently failing on v1.14 branch: https://github.com/cilium/cilium/actions/workflows/tests-ipsec-upgrade.yaml?query=branch%3Av1.14

Test: Setup & Test (2, 5.10-20231026.065108, iptables, disabled, disabled, ipsec, true)

```
Test [node-to-node-encryption]:
  ❌ node-to-node-encryption/node-to-node-encryption/ping-ipv4: cilium-test/client-75bff5f5b9-8qpdt (10.244.2.88) -> cilium-test/host-netns-nrpzt (172.18.0.2:0)
  ❌ node-to-node-encryption/node-to-node-encryption/ping-ipv6: cilium-test/client-75bff5f5b9-8qpdt (fd00:10:244:2::59c3) -> cilium-test/host-netns-nrpzt (fc00:c111::2:0)
connectivity test failed: 1 tests failed
```

Example run: https://github.com/cilium/cilium/actions/runs/6961101017/job/18942025818
It was enabled recently in https://github.com/cilium/cilium/pull/29225

**Comment by qmonnet** [comment]
Some observations

### Context

The issue occurs on branches `v1.13` and `v1.14`, but seemingly not on `main`.

It happens on the jobs running with kernels 5.4 (sometimes) or 5.10 (more often). The 5.10 job has `endpoint-routes` enabled, the 5.4 job has not. The failure does not seem to occur with bpf-next (no `endpoint-routes`, but VXLAN tunnelling when the other jobs disable tunnels). I don't know for the newer 6.1 job, haven't observed on it yet.

It can happen after upgrade or after downgrade, be it from/to minor or patch versions.

It happens whether IPsec is enabled or not (I tried disabling it).

It can be reproduced locally, although just like on CI, it does not fail all the time.

<details><summary>Steps</summary>

On top of `v1.13`

```bash
sudo sysctl fs.inotify.max_user_watches=524288
sudo sysctl fs.inotify.max_user_instances=512
if [ ! -d "cilium-v1.12" ]; then
    git clone --depth 1 --branch v1.12 file:///home/qmo/dev/cilium cilium-v1.12
fi

./contrib/scripts/kind.sh "" 3 "" "" "iptables" "dual"
kubectl patch node kind-worker3 --type=json -p='[{"op":"add","path":"/metadata/labels/cilium.io~1no-schedule","value":"true"}]'
kubectl create -n kube-system secret generic cilium-ipsec-keys \
    --from-literal=keys="3 rfc4106(gcm(aes)) $(echo $(dd if=/dev/urandom count=20 bs=1 2> /dev/null | xxd -p -c 64)) 128"

./cilium-cli install \
    --wait \
    --chart-directory=./cilium-v1.12/install/kubernetes/cilium/ \
    --helm-set=debug.enabled=true \
    --helm-set=debug.verbose=envoy \
    --helm-set=hubble.eventBufferCapacity=65535 \
    --helm-set=bpf.monitorAggregation=none \
    --helm-set=cluster.name=default \
    --nodes-without-cilium=kind-worker3 \
    --helm-set-string=kubeProxyReplacement=disabled \
    --set='bpfClockProbe=false,cni.uninstall=false' \
    --helm-set=image.repository=quay.io/cilium/cilium-ci \
    --helm-set=image.useDigest=false \
    --helm-set=image.tag=v1.12 \
    --helm-set=operator.image.repository=quay.io/cilium/operator \
    --helm-set=operator.image.suffix=-ci \
    --helm-set=operator.image.tag=v1.12 \
    --helm-set=operator.image.useDigest=false \
    --helm-set=hubble.relay.image.repository=quay.io/cilium/hubble-relay-ci \
    --helm-set=hubble.relay.image.tag=v1.12 \
    --helm-set=hubble.relay.image.useDigest=false \
    --helm-set-string=tunnel=disabled \
    --helm-set-string=autoDirectNodeRoutes=true \
    --helm-set-string=ipv4NativeRoutingCIDR=10.244.0.0/16 \
    --helm-set-string=tunnel=disabled \
    --helm-set-string=ipv6NativeRoutingCIDR=fd00:10:244::/56 \
    --helm-set-string=endpointRoutes.enabled=true \
    --helm-set=encryption.enabled=true \
    --helm-set=encryption.type=ipsec

./cilium-cli connectivity test --include-conn-disrupt-test --conn-disrupt-test-setup

./cilium-cli upgrade \
    --wait \
    --chart-directory=./install/kubernetes/cilium \
    --helm-set=debug.enabled=true \
    --helm-set=debug.verbose=envoy \
    --helm-set=hubble.eventBufferCapacity=65535 \
    --helm-set=bpf.monitorAggregation=none \
    --helm-set=cluster.name=default \
    --nodes-without-cilium=kind-worker3 \
    --helm-set-string=kubeProxyReplacement=disabled \
    --set='bpfClockProbe=false,cni.uninstall=false' \
    --helm-set=image.repository=quay.io/cilium/cilium-ci \
    --helm-set=image.useDigest=false \
    --helm-set=image.tag=b60f41af8bc6782535e636e108d1e97ae5a4f657 \
    --helm-set=operator.image.repository=quay.io/cilium/operator \
    --helm-set=operator.image.suffix=-ci \
    --helm-set=operator.image.tag=b60f41af8bc6782535e636e108d1e97ae5a4f657 \
    --helm-set=operator.image.useDigest=false \
    --helm-set=hubble.relay.image.repository=quay.io/cilium/hubble-relay-ci \
    --helm-set=hubble.relay.image.tag=b60f41af8bc6782535e636e108d1e97ae5a4f657 \
    --helm-set=hubble.relay.image.useDigest=false \
    --helm-set-string=tunnel=disabled \
    --helm-set-string=autoDirectNodeRoutes=true \
    --helm-set-string=ipv4NativeRoutingCIDR=10.244.0.0/16 \
    --helm-set-string=tunnel=disabled \
    --helm-set-string=ipv6NativeRoutingCIDR=fd00:10:244::/56 \
    --helm-set-string=endpointRoutes.enabled=true \
    --helm-set=encryption.enabled=true \
    --helm-set=encryption.type=ipsec

./cilium-cli connectivity test --include-conn-disrupt-test \
    --include-unsafe-tests \
    --debug \
    --flush-ct

./cilium-cli connectivity test --include-conn-disrupt-test --conn-disrupt-test-setup

./cilium-cli upgrade \
    --wait \
    --chart-directory=./cilium-v1.12/install/kubernetes/cilium/ \
    --helm-set=debug.enabled=true \
    --helm-set=debug.verbose=envoy \
    --helm-set=hubble.eventBufferCapacity=65535 \
    --helm-set=bpf.monitorAggregation=none \
    --helm-set=cluster.name=default \
    --nodes-without-cilium=kind-worker3 \
    --helm-set-string=kubeProxyReplacement=disabled \
    --set='bpfClockProbe=false,cni.uninstall=false' \
    --helm-set=image.repository=quay.io/cilium/cilium-ci \
    --helm-set=image.useDigest=false \
    --helm-set=image.tag=v1.12 \
    --helm-set=operator.image.repository=quay.io/cilium/operator \
    --helm-set=operator.image.suffix=-ci \
    --helm-set=operator.image.tag=v1.12 \
    --helm-set=operator.image.useDigest=false \
    --helm-set=hubble.relay.image.repository=quay.io/cilium/hubble-relay-ci \
    --helm-set=hubble.relay.image.tag=v1.12 \
    --helm-set=hubble.relay.image.useDigest=false \
    --helm-set-string=tunnel=disabled \
    --helm-set-string=autoDirectNodeRoutes=true \
    --helm-set-string=ipv4NativeRoutingCIDR=10.244.0.0/16 \
    --helm-set-string=tunnel=disabled \
    --helm-set-string=ipv6NativeRoutingCIDR=fd00:10:244::/56 \
    --helm-set-string=endpointRoutes.enabled=true \
    --helm-set=encryption.enabled=true \
    --helm-set=encryption.type=ipsec

./cilium-cli connectivity test --include-conn-disrupt-test \
    --include-unsafe-tests \
    --debug \
    --flush-ct
```
</details>

### Failure

When the failure occurs, we get a message such as the one pasted by Marcel:

```
Test [node-to-node-encryption]:
  ❌ node-to-node-encryption/node-to-node-encryption/ping-ipv4: cilium-test/client-75bff5f5b9-8qpdt (10.244.2.88) -> cilium-test/host-netns-nrpzt (172.18.0.2:0)
```

Looking at the debug output from `cilium-cli`, we see that `cilium-test/client-75bff5f5b9-8qpdt` pinged `cilium-test/host-netns-nrpzt`, and that a `tcpdump` session on the client's node (on the `host-netns` pod of the client's node) failed to observe the expected unencrypted packets.

We fail to observe these packets with `tcpdump` because we expect packets with the pod IP as source address (`tcpdump <options> host src 10.244.2.88 and icmp and host dst 172.18.0.2`). However, the packets reach the interface with the node IP instead (e.g. `172.18.0.3`).

I have not found why this masquerading is happening yet, let alone why this only occurs occasionally.

I have not managed to reach the failure state locally without running the full suite of connectivity tests, (but once the failure state is reach, subsequent runs of `node-to-node-encryption` test all fail), so I don't rule out a side effect from one of the previous tests. However, I have not managed to reach the failure state locally many times, so I can't be sure it's related to the previous tests, either.

**Comment by qmonnet** [comment]
Sample from `pwru` in the case of failure:

```
              ip_forward 10.244.3.36:0->172.19.0.4:0(icmp)
     __xfrm_policy_check 10.244.3.36:0->172.19.0.4:0(icmp)
         decode_session4 10.244.3.36:0->172.19.0.4:0(icmp)
security_xfrm_decode_session 10.244.3.36:0->172.19.0.4:0(i
    __xfrm_route_forward 10.244.3.36:0->172.19.0.4:0(icmp)
         decode_session4 10.244.3.36:0->172.19.0.4:0(icmp)
security_xfrm_decode_session 10.244.3.36:0->172.19.0.4:0(i
            nf_hook_slow 10.244.3.36:0->172.19.0.4:0(icmp)
       ip_forward_finish 10.244.3.36:0->172.19.0.4:0(icmp)
               ip_output 10.244.3.36:0->172.19.0.4:0(icmp)
            nf_hook_slow 10.244.3.36:0->172.19.0.4:0(icmp)
   apparmor_ip_postroute 10.244.3.36:0->172.19.0.4:0(icmp)
     skb_ensure_writable 10.244.3.36:0->172.19.0.4:0(icmp)
     skb_ensure_writable 10.244.3.36:0->172.19.0.4:0(icmp)
inet_proto_csum_replace4 10.244.3.36:0->172.19.0.4:0(icmp)
   __xfrm_decode_session 172.19.0.3:0->172.19.0.4:0(icmp)
         decode_session4 172.19.0.3:0->172.19.0.4:0(icmp)
security_xfrm_decode_session 172.19.0.3:0->172.19.0.4:0(ic
        ip_finish_output 172.19.0.3:0->172.19.0.4:0(icmp)
      __ip_finish_output 172.19.0.3:0->172.19.0.4:0(icmp)
       ip_finish_output2 172.19.0.3:0->172.19.0.4:0(icmp)
        __dev_queue_xmit 172.19.0.3:0->172.19.0.4:0(icmp)
```

Sample of `pwru` in the case of success:

```
              ip_forward 10.244.3.36:0->172.19.0.4:0(icmp)
     __xfrm_policy_check 10.244.3.36:0->172.19.0.4:0(icmp)
         decode_session4 10.244.3.36:0->172.19.0.4:0(icmp)
security_xfrm_decode_session 10.244.3.36:0->172.19.0.4:0(icmp)
    __xfrm_route_forward 10.244.3.36:0->172.19.0.4:0(icmp)
         decode_session4 10.244.3.36:0->172.19.0.4:0(icmp)
security_xfrm_decode_session 10.244.3.36:0->172.19.0.4:0(icmp)
            nf_hook_slow 10.244.3.36:0->172.19.0.4:0(icmp)
       ip_forward_finish 10.244.3.36:0->172.19.0.4:0(icmp)
               ip_output 10.244.3.36:0->172.19.0.4:0(icmp)
            nf_hook_slow 10.244.3.36:0->172.19.0.4:0(icmp)
   apparmor_ip_postroute 10.244.3.36:0->172.19.0.4:0(icmp)
                                                           <- no skb_ensure_writable
                                                           <- no skb_ensure_writable
                                                           <- no inet_proto_csum_replace4
                                                           <- no __xfrm_decode_session
                                                           <- no decode_session4
                                                           <- no security_xfrm_decode_session
        ip_finish_output 10.244.3.36:0->172.19.0.4:0(icmp)
      __ip_finish_output 10.244.3.36:0->172.19.0.4:0(icmp)
       ip_finish_output2 10.244.3.36:0->172.19.0.4:0(icmp)
        __dev_queue_xmit 10.244.3.36:0->172.19.0.4:0(icmp)
```

Stack for `inet_proto_csum_replace4` (failure case), note that the function is called from the `nf_nat` which is not called when the test succeeds. This is where the masquerading occurs (note the difference of IP before/after).

```
0xffffa09a81447a00      5 [<empty>(369538)] inet_proto_csum_replace4 10.244.3.36:0->172.19.0.4:0(icmp)                                                                   
inet_proto_csum_replace4                                                                                                                                                 
nf_nat_ipv4_manip_pkt   [nf_nat]                                                                                                                                         
nf_nat_manip_pkt        [nf_nat]                                                                                                                                         
nf_nat_inet_fn  [nf_nat]                                                                                                                                                 
nf_nat_ipv4_out [nf_nat]                                                                                                                                                 
nf_hook_slow                                                                                                                                                             
ip_output                                                                                                                                                                
ip_forward_finish                                                                                                                                                        
ip_forward                                                                                                                                                               
ip_rcv                                                                                                                                                                   
__netif_receive_skb_one_core                                                                                                                                             
__netif_receive_skb                                                                                                                                                      
process_backlog                                                                                                                                                          
__napi_poll                                                                                                                                                              
net_rx_action                                                                                                                                                            
__do_softirq                                                                                                                                                             
do_softirq.part.0                                                                                                                                                        
__local_bh_enable_ip
ip_finish_output2
__ip_finish_output
ip_finish_output
ip_output
ip_push_pending_frames
raw_sendmsg
inet_sendmsg
sock_sendmsg
__sys_sendto
__x64_sys_sendto
do_syscall_64
entry_SYSCALL_64_after_hwframe
0xffffa09a81447a00      5 [<empty>(369538)]    __xfrm_decode_session 172.19.0.3:0->172.19.0.4:0(icmp)
```

Looking at `ip_output()`'s code, I don't understand why we call the `nf_nat` hook in some cases, but not others. I do see a condition to validate (`!(IPCB(skb)->flags & IPSKB_REROUTED)`), but as I read it this relates to `skb.cb`, and we have `skb.cb = (char[])[9,]` in both cases, says `pwru`. And I suppose the hooks should exist in both cases, given that we have the same iptables rules. :thinking:

**Comment by qmonnet** [comment]
As mentioned earlier I could also reproduce with encryption **disabled**, but this time I also looked at the output from `pwru`. In case of failure we get mostly the same path (we skip `__xfrm_policy_check` but still take the `nf_nat` hook in `ip_output()` and masquerade the address right before `__xfrm_decode_session()`, which I'm not even sure why we call it given that encryption is off). 

Sample `pwru` output in case of test failure, no encryption

```
              ip_forward 10.244.2.162:0->172.19.0.2:0(icmp)
            nf_hook_slow 10.244.2.162:0->172.19.0.2:0(icmp)   <- we skip some xfrm stuff before this
       ip_forward_finish 10.244.2.162:0->172.19.0.2:0(icmp)   <- same as with encryption from here
               ip_output 10.244.2.162:0->172.19.0.2:0(icmp)
            nf_hook_slow 10.244.2.162:0->172.19.0.2:0(icmp)
   apparmor_ip_postroute 10.244.2.162:0->172.19.0.2:0(icmp)
     skb_ensure_writable 10.244.2.162:0->172.19.0.2:0(icmp)
     skb_ensure_writable 10.244.2.162:0->172.19.0.2:0(icmp)
inet_proto_csum_replace4 10.244.2.162:0->172.19.0.2:0(icmp)
   __xfrm_decode_session 172.19.0.3:0->172.19.0.2:0(icmp)     <- source IP changes here
         decode_session4 172.19.0.3:0->172.19.0.2:0(icmp)
security_xfrm_decode_session 172.19.0.3:0->172.19.0.2:0(icmp)
        ip_finish_output 172.19.0.3:0->172.19.0.2:0(icmp)
      __ip_finish_output 172.19.0.3:0->172.19.0.2:0(icmp)
       ip_finish_output2 172.19.0.3:0->172.19.0.2:0(icmp)
    neigh_resolve_output 172.19.0.3:0->172.19.0.2:0(icmp)
              eth_header 172.19.0.3:0->172.19.0.2:0(icmp)
                skb_push 172.19.0.3:0->172.19.0.2:0(icmp)
        __dev_queue_xmit 172.19.0.3:0->172.19.0.2:0(icmp)
```

**Comment by qmonnet** [comment]
> I wonder if `pwru` might fail to display `cb[]` properly, given that it's an array (could it be printing the first value only, and would we be missing the actual flag values?). This would correspond to the success case entering `NF_HOOK_COND()` in `ip_output()`, failing the condition? then directly calling `ip_finish_output()`. This would also mean that we would have an inconsistency in the `skb->dst_entry->xfrm` (`dst` as in _destination cache definitions_), which is what triggers `IPSKB_REROUTED` to be set.

It seems that pwru does have a bug for printing `skb->cb[]`. I hacked something (thanks Martynas!) in it to dump the full `cb[]`, and I see more values:

```
00:     0b 00 00 00 00 00 00 00   <- first 4 bytes are iif, here 11
08:     00 00 00 00 00 00 00 00
16:     00 00 00 00 01 00 00 00
24:     00 00 00 00 00 00 00 00
32:     <some 8-bytes pointer>
40:     <some 8-bytes pointer>
```

From what I read, `IPCB(skb)` casts `cb` to a `struct inet_skb_parm` where 4 first bytes are `iif`, followed by a 16-char wide `struct ip_options`, and then come the 2-byte long `flags`. This gives us `flags == 1` (`IPSKB_FORWARDED`), if I'm not mistaken. We don't have the `IPSKB_REROUTED` set.

This is what I expected for the test failure case. I hoped to observe something different when the test succeeds; alas, the `flags` are exactly the same, so this doesn't seem to explain why we sometimes end up going through the `nf_nat` hook and sometimes not :disappointed:.

**Comment by brb** [comment]
@qmonnet Great investigation! Maybe you have pwru logs w/ `--output-meta`? That could help us track down which ifaces the troublesome pkt traversed. Maybe it will give us more clue :thinking:

**Comment by qmonnet** [comment]
So I think the `ipset` entry is missing because we sometimes remove it by accident.

On line `https://github.com/cilium/cilium/blob/bae5aecfd0ded9e77e56a5123350494386576cb3/pkg/node/manager/manager.go#L531` (current v1.13 HEAD), in `NodeUpdated()`, we look for an item in a slice as part of the following block:

```go
	if option.Config.NodeIpsetNeeded() && address.Type == addressing.NodeInternalIP &&
		!slices.Contains(ipsAdded, address.IP.String()) {
		iptables.RemoveFromNodeIpset(address.IP)
	}
```

But the search in `slices.Contains(ipsAdded, address.IP.String()` does not use the correct value. When we add an entry to the `ipsAdded` slice, we use a value of type `netip.Prefix` converted with `.String()`, and this results in an IP _with a CIDR mask (using the slash notation)_. Meanwhile, `address.IP.String()` is a mere IP, with no CIDR mask attached. For example, during my tests, I've observed the slice `ipsAdded` containing: `[172.19.0.2/32 fc00:c111::2/128]`, and the `address.IP.String()` being `172.19.0.2` (no CIDR mask). As a consequence of that, the search fails, the `if` condition passes, and we do remove the entry from the ipset, even though we just added it above in the function and are supposed to keep it.

I'm currently trying to validate a fix with https://github.com/cilium/cilium/pull/29898.
